File: backend/mr-db-sql/src/main/java/io/featurehub/db/services/ServiceAccountSqlApi.java
package io.featurehub.db.services;

import io.ebean.Database;
import io.ebean.DuplicateKeyException;
import io.ebean.annotation.Transactional;
import io.featurehub.db.api.FillOpts;
import io.featurehub.db.api.OptimisticLockingException;
import io.featurehub.db.api.Opts;
import io.featurehub.db.api.ServiceAccountApi;
import io.featurehub.db.model.DbAcl;
import io.featurehub.db.model.DbApplication;
import io.featurehub.db.model.DbEnvironment;
import io.featurehub.db.model.DbPerson;
import io.featurehub.db.model.DbPortfolio;
import io.featurehub.db.model.DbServiceAccount;
import io.featurehub.db.model.DbServiceAccountEnvironment;
import io.featurehub.db.model.query.QDbAcl;
import io.featurehub.db.model.query.QDbEnvironment;
import io.featurehub.db.model.query.QDbServiceAccount;
import io.featurehub.db.model.query.QDbServiceAccountEnvironment;
import io.featurehub.db.publish.CacheSource;
import io.featurehub.mr.model.Person;
import io.featurehub.mr.model.PublishAction;
import io.featurehub.mr.model.RoleType;
import io.featurehub.mr.model.ServiceAccount;
import io.featurehub.mr.model.ServiceAccountPermission;
import org.apache.commons.lang3.RandomStringUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.inject.Inject;
import javax.inject.Singleton;
import java.util.ArrayList;
import java.util.Collection;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.Set;
import java.util.UUID;
import java.util.function.Function;
import java.util.stream.Collectors;

@Singleton
public class ServiceAccountSqlApi implements ServiceAccountApi {
  private static final Logger log = LoggerFactory.getLogger(ServiceAccountSqlApi.class);
  private final Database database;
  private final Conversions convertUtils;
  private final CacheSource cacheSource;
  private final ArchiveStrategy archiveStrategy;

  @Inject
  public ServiceAccountSqlApi(
      Database database,
      Conversions convertUtils,
      CacheSource cacheSource,
      ArchiveStrategy archiveStrategy) {
    this.database = database;
    this.convertUtils = convertUtils;
    this.cacheSource = cacheSource;
    this.archiveStrategy = archiveStrategy;
  }

  @Override
  public ServiceAccount get(UUID id, Opts opts) {
    Conversions.nonNullServiceAccountId(id);

    QDbServiceAccount eq = opts(new QDbServiceAccount().id.eq(id), opts);
    return convertUtils.toServiceAccount(eq.findOne(), opts);
  }

  private QDbServiceAccount opts(QDbServiceAccount finder, Opts opts) {
    if (opts.contains(FillOpts.Permissions) || opts.contains(FillOpts.SdkURL)) {
      finder = finder.serviceAccountEnvironments.fetch();
    }
    if (!opts.contains(FillOpts.Archived)) {
      finder = finder.whenArchived.isNull();
    }
    return finder;
  }

  @Override
  public ServiceAccount update(UUID id, Person updater, ServiceAccount serviceAccount, Opts opts)
      throws OptimisticLockingException {
    Conversions.nonNullServiceAccountId(id);

    DbServiceAccount sa = new QDbServiceAccount().id.eq(id).whenArchived.isNull().findOne();

    if (sa == null) return null;

    if (serviceAccount.getVersion() == null || serviceAccount.getVersion() != sa.getVersion()) {
      throw new OptimisticLockingException();
    }

    Map<UUID, ServiceAccountPermission> updatedEnvironments = new HashMap<>();
    List<UUID> newEnvironments = new ArrayList<>();

    serviceAccount
        .getPermissions()
        .forEach(
            perm -> {
              if (perm.getEnvironmentId() != null) {
                updatedEnvironments.put(perm.getEnvironmentId(), perm);
                newEnvironments.add(perm.getEnvironmentId());
              }
            });

    List<DbServiceAccountEnvironment> deletePerms = new ArrayList<>();
    List<DbServiceAccountEnvironment> updatePerms = new ArrayList<>();
    List<DbServiceAccountEnvironment> createPerms = new ArrayList<>();

    // we drop out of this knowing which perms to delete and update
    new QDbServiceAccountEnvironment()
        .environment
        .id
        .in(updatedEnvironments.keySet())
        .serviceAccount
        .eq(sa)
        .findEach(
            upd -> {
              final UUID envId = upd.getEnvironment().getId();

              final ServiceAccountPermission perm = updatedEnvironments.get(envId);

              newEnvironments.remove(envId);

              if (perm.getPermissions() == null || perm.getPermissions().isEmpty()) {
                deletePerms.add(upd);
              } else {
                final String newPerms = convertPermissionsToString(perm.getPermissions());
                if (!newPerms.equals(upd.getPermissions())) {
                  upd.setPermissions(newPerms);
                  updatePerms.add(upd);
                }
              }
            });

    // now we need to know which perms to add
    newEnvironments.forEach(
        envId -> {
          final ServiceAccountPermission perm = updatedEnvironments.get(envId);
          if (perm.getPermissions() != null && !perm.getPermissions().isEmpty()) {
            DbEnvironment env =
                convertUtils.byEnvironment(
                    envId, Opts.opts(FillOpts.ApplicationIds, FillOpts.PortfolioIds));

            if (env != null
                && env.getParentApplication()
                    .getPortfolio()
                    .getId()
                    .equals(sa.getPortfolio().getId())) {
              createPerms.add(
                  new DbServiceAccountEnvironment.Builder()
                      .environment(env)
                      .serviceAccount(sa)
                      .permissions(convertPermissionsToString(perm.getPermissions()))
                      .build());
            }
          }
        });

    if (serviceAccount.getDescription() != null) {
      sa.setDescription(serviceAccount.getDescription());
    }

    updateServiceAccount(sa, deletePerms, updatePerms, createPerms);

    return convertUtils.toServiceAccount(sa, opts);
  }

  private String convertPermissionsToString(List<RoleType> permissions) {
    return permissions.stream().map(RoleType::toString).sorted().collect(Collectors.joining(","));
  }

  @Override
  public List<ServiceAccount> search(
      UUID pId, String filter, UUID applicationId, Person currentPerson, Opts opts) {
    Conversions.nonNullPortfolioId(pId);

    DbPerson person = convertUtils.byPerson(currentPerson);

    if (person == null) {
      return null;
    }

    DbApplication application = null;
    boolean personAdmin = false;
    if (applicationId != null) {
      application = convertUtils.byApplication(applicationId);

      if (application != null) {
        personAdmin = convertUtils.isPersonApplicationAdmin(person, application);
      }
    }

    QDbServiceAccount qFinder = opts(new QDbServiceAccount().portfolio.id.eq(pId), opts);
    if (filter != null) {
      qFinder = qFinder.name.ilike(filter);
    }

    if (application != null) {
      qFinder = qFinder.portfolio.applications.eq(application);

      qFinder = opts(qFinder, opts);
    }

    final Opts updatedOpts = opts;

    // if they are an admin they have everything, otherwise spelunk through finding relevant ACLs
    final List<DbAcl> environmentPermissions = new ArrayList<>();

    if (!personAdmin
        && application != null
        && (opts.contains(FillOpts.Permissions) || opts.contains(FillOpts.SdkURL))) {
      environmentPermissions.addAll(
          new QDbAcl()
              .roles
              .notEqualTo("")
              .environment
              .parentApplication
              .eq(application)
              .group
              .peopleInGroup
              .eq(person)
              .environment
              .fetch(QDbEnvironment.Alias.id)
              .findList());
    }

    return qFinder.findList().stream()
        .map(sa -> convertUtils.toServiceAccount(sa, updatedOpts, environmentPermissions))
        .collect(Collectors.toList());
  }

  @Override
  public ServiceAccount resetApiKey(UUID id) {
    Conversions.nonNullServiceAccountId(id);

    DbServiceAccount sa = new QDbServiceAccount().id.eq(id).whenArchived.isNull().findOne();
    if (sa == null) return null;

    sa.setApiKeyServerEval(newServerEvalKey());
    sa.setApiKeyClientEval(newClientEvalKey());
    updateOnlyServiceAccount(sa);
    asyncUpdateCache(sa, null);
    return convertUtils.toServiceAccount(sa, Opts.empty());
  }

  @Transactional
  public void cleanupServiceAccountApiKeys() {
    if (new QDbServiceAccount()
            .or()
            .apiKeyClientEval
            .isNull()
            .apiKeyServerEval
            .isNull()
            .endOr()
            .exists()) {
      log.info("Updating service account keys as incomplete.");
      new QDbServiceAccount()
          .or()
          .apiKeyClientEval
          .isNull()
          .apiKeyServerEval
          .isNull()
          .endOr()
          .findEach(
              sa -> {
                boolean updated = false;
                if (sa.getApiKeyClientEval() == null) {
                  updated = true;
                  sa.setApiKeyClientEval(newClientEvalKey());
                }
                if (sa.getApiKeyServerEval() == null) {
                  updated = true;
                  sa.setApiKeyServerEval(newServerEvalKey());
                }
                if (updated) {
                  database.update(sa);
                }
              });
    }
  }

  private String newServerEvalKey() {
    return RandomStringUtils.randomAlphanumeric(40);
  }

  private String newClientEvalKey() {
    return RandomStringUtils.randomAlphanumeric(30)
        + "*"
        + RandomStringUtils.randomAlphanumeric(20);
  }

  @Override
  public ServiceAccount create(
      UUID portfolioId, Person creator, ServiceAccount serviceAccount, Opts opts)
      throws DuplicateServiceAccountException {
    Conversions.nonNullPortfolioId(portfolioId);
    Conversions.nonNullPerson(creator);

    DbPerson who = convertUtils.byPerson(creator);
    DbPortfolio portfolio = convertUtils.byPortfolio(portfolioId);

    if (who == null || portfolio == null) return null;

    List<DbEnvironment> changedEnvironments = new ArrayList<>();
    Map<UUID, DbEnvironment> envs = environmentMap(serviceAccount);

    // now where we actually find the environment, add it into the list
    Set<DbServiceAccountEnvironment> perms =
        serviceAccount.getPermissions().stream()
            .map(
                sap -> {
                  if (sap.getEnvironmentId() != null) {
                    DbEnvironment e = envs.get(sap.getEnvironmentId());
                    if (e != null) {
                      changedEnvironments.add(e);
                      return new DbServiceAccountEnvironment.Builder()
                          .environment(e)
                          .permissions(convertPermissionsToString(sap.getPermissions()))
                          .build();
                    }
                  }
                  return null;
                })
            .filter(Objects::nonNull)
            .collect(Collectors.toSet());

    // now create the SA and attach the perms to form the links
    DbServiceAccount sa =
        new DbServiceAccount.Builder()
            .name(serviceAccount.getName())
            .description(serviceAccount.getDescription())
            .whoChanged(who)
            .apiKeyServerEval(newServerEvalKey())
            .apiKeyClientEval(newClientEvalKey())
            .serviceAccountEnvironments(perms)
            .portfolio(portfolio)
            .build();

    perms.forEach(p -> p.setServiceAccount(sa));

    try {
      save(sa);

      asyncUpdateCache(sa, changedEnvironments);
    } catch (DuplicateKeyException dke) {
      log.warn("Duplicate service account {}", sa.getName(), dke);
      throw new DuplicateServiceAccountException();
    }

    return convertUtils.toServiceAccount(sa, opts);
  }

  private Map<UUID, DbEnvironment> environmentMap(ServiceAccount serviceAccount) {
    // find all of the UUIDs in the environment list
    List<UUID> envIds =
        serviceAccount.getPermissions().stream()
            .map(ServiceAccountPermission::getEnvironmentId)
            .filter(Objects::nonNull)
            .collect(Collectors.toList());

    // now find them in the db in one swoop using "in" syntax
    return new QDbEnvironment()
        .id.in(envIds).whenArchived.isNull().findList().stream()
            .collect(Collectors.toMap(DbEnvironment::getId, Function.identity()));
  }

  @Transactional
  private void save(DbServiceAccount sa) {
    database.save(sa);
  }

  @Transactional
  private void updateOnlyServiceAccount(DbServiceAccount sa) {
    database.update(sa);
  }

  @Transactional
  private void updateServiceAccount(
      DbServiceAccount sa,
      List<DbServiceAccountEnvironment> deleted,
      List<DbServiceAccountEnvironment> updated,
      List<DbServiceAccountEnvironment> created) {
    database.update(sa);

    database.updateAll(updated);
    database.deleteAll(deleted);
    database.saveAll(created);

    Map<UUID, DbEnvironment> changed = new HashMap<>();

    deleted.forEach(e -> changed.put(e.getEnvironment().getId(), e.getEnvironment()));
    updated.forEach(e -> changed.put(e.getEnvironment().getId(), e.getEnvironment()));
    created.forEach(e -> changed.put(e.getEnvironment().getId(), e.getEnvironment()));

    asyncUpdateCache(sa, changed.values());
  }

  // because this is an update or save, its no problem we send this out of band of this save/update.
  private void asyncUpdateCache(
      DbServiceAccount sa, Collection<DbEnvironment> changedEnvironments) {
    cacheSource.updateServiceAccount(sa, PublishAction.UPDATE);

    if (changedEnvironments != null && !changedEnvironments.isEmpty()) {
      changedEnvironments.forEach(e -> cacheSource.updateEnvironment(e, PublishAction.UPDATE));
    }
  }

  @Override
  @Transactional
  public Boolean delete(Person deleter, UUID servcieAccountId) {
    Conversions.nonNullServiceAccountId(servcieAccountId);

    DbServiceAccount sa =
        new QDbServiceAccount().id.eq(servcieAccountId).whenArchived.isNull().findOne();
    if (sa != null) {
      archiveStrategy.archiveServiceAccount(sa);
      return Boolean.TRUE;
    }

    return Boolean.FALSE;
  }
}

File: backend/mr-db-sql/src/main/kotlin/io/featurehub/db/services/ServiceAccountSqlApi.kt
package io.featurehub.db.services

import io.ebean.DuplicateKeyException
import io.ebean.annotation.Transactional
import io.ebean.annotation.TxType
import io.featurehub.dacha.model.PublishAction
import io.featurehub.db.api.FillOpts
import io.featurehub.db.api.OptimisticLockingException
import io.featurehub.db.api.Opts
import io.featurehub.db.api.ServiceAccountApi
import io.featurehub.db.model.*
import io.featurehub.db.model.query.QDbAcl
import io.featurehub.db.model.query.QDbEnvironment
import io.featurehub.db.model.query.QDbFeatureFilter
import io.featurehub.db.model.query.QDbServiceAccount
import io.featurehub.db.model.query.QDbServiceAccountEnvironment
import io.featurehub.mr.events.common.CacheSource
import io.featurehub.mr.model.*
import io.featurehub.mr.model.RoleType
import jakarta.inject.Inject
import jakarta.inject.Singleton
import org.apache.commons.lang3.RandomStringUtils
import org.slf4j.LoggerFactory
import java.time.Instant
import java.util.*

@Singleton
class ServiceAccountSqlApi @Inject constructor(
  private val convertUtils: Conversions,
  private val cacheSource: CacheSource,
  private val archiveStrategy: ArchiveStrategy,
  private val internalPersonApi: InternalPersonApi
) : ServiceAccountApi, InternalServiceAccountApi {
  override fun get(id: UUID, opts: Opts): ServiceAccount? {
    val eq = opts(QDbServiceAccount().id.eq(id), opts)
    return convertUtils.toServiceAccount(eq.findOne(), opts)
  }

  private fun opts(finder: QDbServiceAccount, opts: Opts?): QDbServiceAccount {
    var qFinder = finder
    opts?.let { opt ->
      if (opt.contains(FillOpts.Permissions) || opt.contains(FillOpts.SdkURL)) {
        qFinder = qFinder.serviceAccountEnvironments.fetch()
      }

      if (opt.contains(FillOpts.ServiceAccountFilters)) {
        qFinder = qFinder.featureFilters.fetch()
      }

      if (!opt.contains(FillOpts.Archived)) {
        qFinder = qFinder.whenArchived.isNull
      }
    }
    return qFinder
  }

  @Throws(OptimisticLockingException::class)
  override fun update(portfolioId: UUID, personId: UUID, serviceAccount: ServiceAccount, appId: UUID?, opts: Opts): ServiceAccount? {
    val sa = QDbServiceAccount().id.eq(serviceAccount.id).whenArchived.isNull.portfolio.id.eq(portfolioId).findOne() ?: return null

    if (serviceAccount.version == null || serviceAccount.version != sa.version) {
      throw OptimisticLockingException()
    }

    val whoUpdated = convertUtils.byPerson(personId) ?: return null
    return update(sa, whoUpdated, serviceAccount, appId, opts, portfolioId)
  }

  @Throws(OptimisticLockingException::class)
  override fun update(
    serviceAccountId: UUID,
    updater: Person,
    serviceAccount: ServiceAccount,
    appId: UUID?,
    opts: Opts,
    portfolioId: UUID
  ): ServiceAccount? {
    val sa = QDbServiceAccount().id.eq(serviceAccountId).whenArchived.isNull.findOne() ?: return null

    if (serviceAccount.version == null || serviceAccount.version != sa.version) {
      throw OptimisticLockingException()
    }
    val whoUpdated = convertUtils.byPerson(updater) ?: return null
    return update(sa, whoUpdated, serviceAccount, appId, opts, portfolioId)
  }

  private fun update(
    sa: DbServiceAccount,
    whoUpdated: DbPerson,
    serviceAccount: ServiceAccount,
    appId: UUID?,
    opts: Opts,
    portfolioId: UUID
  ): ServiceAccount {
    val permChanges = ServiceAccountPermsUpdated()

    if (serviceAccount.permsInvalid == null || serviceAccount.permsInvalid == false) {
      permChanges.update(serviceAccount, appId, sa, convertUtils)
    }

    var descUpdated = false
    serviceAccount.description?.let {
      sa.description = it
      descUpdated = true
    }

    var updateAssociatedUser = false

    if (serviceAccount.name != sa.name) {
      sa.name = serviceAccount.name
      updateAssociatedUser = true
    }

    val filterUpdated = updateFilterOnServiceAccount(serviceAccount, sa, portfolioId)

    if (descUpdated || updateAssociatedUser || filterUpdated || permChanges.hasChanges()) {
      sa.whoChanged = whoUpdated
    }

    asyncUpdateCache(sa, updateServiceAccount(sa, permChanges).values)

    if (updateAssociatedUser) {
      internalPersonApi.updateSdkServiceAccountUser(sa.sdkPerson.id, whoUpdated, serviceAccount.name)
    }

    return convertUtils.toServiceAccount(sa, opts)!!
  }

  internal class ServiceAccountPermsUpdated {
    val deletePerms = mutableListOf<DbServiceAccountEnvironment>()
    val updatePerms = mutableListOf<DbServiceAccountEnvironment>()
    val createPerms = mutableListOf<DbServiceAccountEnvironment>()

    fun hasChanges() : Boolean {
      return deletePerms.isNotEmpty() || updatePerms.isNotEmpty() || createPerms.isNotEmpty()
    }

    val updated : List<DbServiceAccountEnvironment>
      get() = this.updatePerms
    val deleted : List<DbServiceAccountEnvironment>
      get() = this.deletePerms
    val created : List<DbServiceAccountEnvironment>
      get() = this.createPerms


    fun update(serviceAccount: ServiceAccount, appId: UUID?, sa: DbServiceAccount, convertUtils: Conversions) {
      val updatedEnvironments: MutableMap<UUID, ServiceAccountPermission> = HashMap()
      val newEnvironments: MutableList<UUID> = ArrayList()
      serviceAccount
        .permissions.forEach { perm: ServiceAccountPermission ->
          updatedEnvironments[perm.environmentId] = perm
          newEnvironments.add(perm.environmentId)
        }

      // we drop out of this knowing which perms to delete and update
      var finder = QDbServiceAccountEnvironment().environment.id
        .`in`(updatedEnvironments.keys).serviceAccount
        .eq(sa)

      // limit our area of interest
      if (appId != null) {
        finder = finder.environment.parentApplication.id.eq(appId)
      }

      finder
        .findEach { upd: DbServiceAccountEnvironment ->
          val envId = upd.environment.id
          val perm = updatedEnvironments[envId]
          newEnvironments.remove(envId)
          if (perm == null || perm.permissions.isEmpty()) {
            deletePerms.add(upd)
          } else {
            val newPerms = convertPermissionsToString(perm.permissions)
            if (newPerms != upd.permissions) {
              upd.permissions = newPerms
              updatePerms.add(upd)
            }
          }
        }

      var deleteFinder = QDbServiceAccountEnvironment().environment.id.notIn(updatedEnvironments.keys).serviceAccount.eq(sa)

      if (appId != null) {
        deleteFinder = deleteFinder.environment.parentApplication.id.eq(appId)
      }

      deleteFinder.findEach { toDelete ->
        deletePerms.add(toDelete)
      }

      // now we need to know which perms to add
      newEnvironments.forEach { envId: UUID ->
        val perm = updatedEnvironments[envId]
        if (perm?.permissions?.isEmpty() == false) {
          val env = convertUtils.byEnvironment(
            envId, Opts.opts(FillOpts.ApplicationIds, FillOpts.PortfolioIds)
          )
          if (env != null
            && (env.parentApplication
              .portfolio
              .id
              == sa.portfolio.id)
          ) {
            createPerms.add(
              DbServiceAccountEnvironment.Builder()
                .environment(env)
                .serviceAccount(sa)
                .permissions(convertPermissionsToString(perm.permissions))
                .build()
            )
          }
        }
      }
    }

    private fun convertPermissionsToString(permissions: List<RoleType>): String {
      return permissions.toSet().joinToString(",") { rt ->
        rt.value
      }
    }
  }

  private fun updateFilterOnServiceAccount(serviceAccount: ServiceAccount, sa: DbServiceAccount, portfolioId: UUID): Boolean {
    // Update filter associations from the featureFilters list (null means "don't change", empty list clears)
    var filterUpdated = false

    serviceAccount.featureFilters?.let { updatedFilters ->
      // remove the records that we won't use any longer
      filterUpdated = sa.featureFilters.removeIf { !updatedFilters.contains(it.id) }

      val remainingIds = sa.featureFilters.map { it.id }

      updatedFilters.removeAll(remainingIds)

      if (updatedFilters.isNotEmpty()) {
        filterUpdated = true
        val toAdd = QDbFeatureFilter().id.`in`(updatedFilters).portfolio.id.eq(portfolioId).findList()
        sa.featureFilters.addAll(toAdd)
      }}

    return filterUpdated
  }

  private fun convertPermissionsToString(permissions: List<RoleType>): String {
    return permissions.toSet().joinToString(",") { rt ->
      rt.value
    }
  }

  override fun search(
    portfolioId: UUID,
    filter: String?,
    applicationId: UUID?,
    currentPerson: Person,
    opts: Opts
  ): List<ServiceAccount> {
    val person = convertUtils.byPerson(currentPerson) ?: return listOf()
    var application: DbApplication? = null
    var personAdmin = false
    if (applicationId != null) {
      application = convertUtils.byApplication(applicationId)
      if (application != null) {
        personAdmin = convertUtils.isPersonApplicationAdmin(person, application)
      }
    }
    var qFinder = opts(QDbServiceAccount().portfolio.id.eq(portfolioId), opts)
    if (filter != null) {
      qFinder = qFinder.name.ilike(filter)
    }
    if (application != null) {
      qFinder = qFinder.portfolio.applications.eq(application)
      qFinder = opts(qFinder, opts)
    }

    // if they are an admin they have everything, otherwise spelunk through finding relevant ACLs
    val environmentPermissions: MutableList<DbAcl> = ArrayList()
    if (!personAdmin
      && application != null && (opts.contains(FillOpts.Permissions) || opts.contains(FillOpts.SdkURL))
    ) {
      environmentPermissions.addAll(
        QDbAcl().roles.notEqualTo("")
          .environment.parentApplication.eq(application)
          .environment.whenUnpublished.isNull
          .environment.whenArchived.isNull
          .group.groupMembers.person.eq(person)
          .environment.fetch(QDbEnvironment.Alias.id)
          .findList()
      )
    }
    return qFinder.findList()
      .map { sa: DbServiceAccount? ->
        convertUtils.toServiceAccount(
          sa,
          opts,
          environmentPermissions
        )!!
      }
  }

  override fun resetApiKey(
    id: UUID,
    resetClientEvalApiKey: Boolean,
    resetServerEvalApiKey: Boolean
  ): ServiceAccount? {
    Conversions.nonNullServiceAccountId(id)
    val sa = QDbServiceAccount().id.eq(id).whenArchived.isNull.findOne() ?: return null
    if (resetServerEvalApiKey) {
      sa.apiKeyServerEval = newServerEvalKey()
    }
    if (resetClientEvalApiKey) {
      sa.apiKeyClientEval = newClientEvalKey()
    }
    updateOnlyServiceAccount(sa)
    asyncUpdateCache(sa, null)
    return convertUtils.toServiceAccount(sa, Opts.empty())
  }

  @Transactional(type = TxType.REQUIRES_NEW)
  override fun cleanupServiceAccountApiKeys() {
    if (QDbServiceAccount()
        .or().apiKeyClientEval
        .isNull.apiKeyServerEval
        .isNull
        .endOr()
        .exists()
    ) {
      log.info("Updating service account keys as incomplete.")
      QDbServiceAccount()
        .or().apiKeyClientEval
        .isNull.apiKeyServerEval
        .isNull
        .endOr()
        .findEach { sa: DbServiceAccount ->
          var updated = false
          if (sa.apiKeyClientEval.isEmpty()) {
            updated = true
            sa.apiKeyClientEval = newClientEvalKey()
          }
          if (sa.apiKeyServerEval.isEmpty()) {
            updated = true
            sa.apiKeyServerEval = newServerEvalKey()
          }
          if (updated) {
            sa.update()
          }
        }
    }
  }

  @Transactional(type = TxType.REQUIRES_NEW)
  override fun unpublishServiceAccounts(portfolioId: UUID, serviceAccounts: List<UUID>?): Int {
    var query = QDbServiceAccount().portfolio.id.eq(portfolioId)
    if (serviceAccounts != null) {
      query = query.id.`in`(serviceAccounts)
    }

    query.select(QDbServiceAccount.Alias.id).findList().forEach { sa ->
      cacheSource.deleteServiceAccount(sa.id)
    }

    return query.asUpdate().set("whenUnpublished", Instant.now()).update()
  }

  private fun newServerEvalKey(): String {
    return RandomStringUtils.randomAlphanumeric(40)
  }

  private fun newClientEvalKey(): String {
    return (RandomStringUtils.randomAlphanumeric(30)
      + "*"
      + RandomStringUtils.randomAlphanumeric(20))
  }

  @Throws(ServiceAccountApi.DuplicateServiceAccountException::class)
  @Transactional
  override fun create(
    portfolioId: UUID,
    creator: Person,
    serviceAccount: CreateServiceAccount,
    opts: Opts
  ): ServiceAccount? {
    val who = convertUtils.byPerson(creator) ?: return null
    val portfolio = convertUtils.byPortfolio(portfolioId) ?: return null

    val changedEnvironments = mutableListOf<DbEnvironment>()
    val envs = environmentMap(serviceAccount)

    // now where we actually find the environment, add it into the list
    val perms = serviceAccount.permissions.mapNotNull { sap: ServiceAccountPermission ->
      envs[sap.environmentId]?.let { e ->
        changedEnvironments.add(e)
        DbServiceAccountEnvironment.Builder()
          .environment(e)
          .permissions(convertPermissionsToString(sap.permissions))
          .build()
      }
    }.toMutableSet()

    val sdkPerson = internalPersonApi.createSdkServiceAccountUser(serviceAccount.name, who, false)
    // now create the SA and attach the perms to form the links
    val sa = DbServiceAccount(who, sdkPerson,
          serviceAccount.name, serviceAccount.description ?: "",
          newServerEvalKey(), newClientEvalKey(), portfolio).let {
      it.serviceAccountEnvironments = perms
      val filterIds = serviceAccount.featureFilter
      if (!filterIds.isNullOrEmpty()) {
        it.featureFilters = QDbFeatureFilter().id.`in`(filterIds).portfolio.id.eq(portfolioId).findList()
      }
      it
    }

    perms.forEach { p: DbServiceAccountEnvironment? -> p!!.serviceAccount = sa }

    try {
      save(sa)
      asyncUpdateCache(sa, changedEnvironments)
    } catch (dke: DuplicateKeyException) {
      log.warn("Duplicate service account {}", sa.name, dke)
      throw ServiceAccountApi.DuplicateServiceAccountException()
    }

    return convertUtils.toServiceAccount(sa, if (opts.contains(FillOpts.Permissions)) opts.add(FillOpts.SdkURL) else opts)
  }

  private fun environmentMap(serviceAccount: CreateServiceAccount): Map<UUID, DbEnvironment> {
    // find all the UUIDs in the environment list
    val envIds = serviceAccount.permissions.map { obj: ServiceAccountPermission -> obj.environmentId }

    // now find them in the db in one swoop using "in" syntax
    return QDbEnvironment().id.`in`(envIds).whenArchived.isNull.findList().associateBy { e -> e.id }
  }

  @Transactional(type = TxType.REQUIRES_NEW)
  private fun save(sa: DbServiceAccount) {
    sa.save()
  }

  @Transactional(type = TxType.REQUIRES_NEW)
  private fun updateOnlyServiceAccount(sa: DbServiceAccount) {
    sa.update()
  }

  @Transactional(type = TxType.REQUIRES_NEW)
  private fun updateServiceAccount(
    sa: DbServiceAccount,
    permChanges: ServiceAccountPermsUpdated
  ) : MutableMap<UUID, DbEnvironment> {
    sa.markAsDirty() // ensure version is changed
    sa.update()
    permChanges.updated.forEach { it.update() }
    permChanges.deleted.forEach { it.delete() }
    permChanges.created.forEach { it.save() }
    val changed = mutableMapOf<UUID, DbEnvironment>()
    permChanges.deleted.forEach { e: DbServiceAccountEnvironment -> changed[e.environment.id] = e.environment }
    permChanges.updated.forEach { e: DbServiceAccountEnvironment -> changed[e.environment.id] = e.environment }
    permChanges.created.forEach { e: DbServiceAccountEnvironment -> changed[e.environment.id] = e.environment }
    changed.values.forEach { e ->
      e.markAsDirty()
      e.update()
    }
    return changed
  }

  // because this is an update or save, its no problem we send this out of band of this save/update.
  private fun asyncUpdateCache(
    sa: DbServiceAccount, changedEnvironments: Collection<DbEnvironment>?
  ) {
    cacheSource.updateServiceAccount(sa, PublishAction.UPDATE)
    changedEnvironments?.forEach { e ->
        cacheSource.updateEnvironment(
          e,
          PublishAction.UPDATE
        )
      }
  }

  @Transactional(type = TxType.REQUIRES_NEW)
  override fun delete(deleter: Person, serviceAccountId: UUID): Boolean {
    val sa = QDbServiceAccount().id.eq(serviceAccountId).whenArchived.isNull.findOne()
    if (sa != null) {
      sa.sdkPerson.let {
          internalPersonApi.deleteSdkServiceAccountUser(it.id, convertUtils.byPerson(deleter)!!)
      }

      archiveStrategy.archiveServiceAccount(sa)
      return true
    }
    return false
  }

  companion object {
    private val log = LoggerFactory.getLogger(ServiceAccountSqlApi::class.java)
  }

  /**
   * This is a transitional job that assumes that all the attached people to the sdkUser are invalid and created
   * new SDK style user accounts for them. After the migration that inserts the job, this is in fact TRUE.
   */
  @Transactional
  override fun ensure_service_accounts_have_person() {
    if (convertUtils.hasOrganisation()) {
      cleanupServiceAccountApiKeys()

      val superuserForOrganisation = mutableMapOf<UUID, DbPerson>()

      QDbServiceAccount().findList().forEach { sa ->
        val orgId = sa.portfolio.organization.id
        val superuser = superuserForOrganisation.computeIfAbsent(orgId) { id ->
          internalPersonApi.findSuperUserToBlame(id)
        }
        sa.sdkPerson = internalPersonApi.createSdkServiceAccountUser(sa.name, superuser, sa.whenArchived != null)
        sa.save()
      }
    }
  }

  // allow us to identify the user who created a feature change for instance
  override fun findServiceAccountByUserId(personId: UUID): UUID? {
    return QDbServiceAccount().sdkPerson.id.eq(personId).findOne()?.id
  }
}
